File: server/src/services/crypto.py
import base64
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


class CryptoService:
    def __init__(self, key: str):
        raw = base64.b64decode(key)
        # AESGCM requires 16, 24, or 32 byte keys
        self._key = raw[:32].ljust(32, b"\0")
        self._aesgcm = AESGCM(self._key)

    def encrypt(self, plaintext: str) -> str:
        nonce = os.urandom(12)
        ciphertext = self._aesgcm.encrypt(nonce, plaintext.encode(), None)
        return base64.b64encode(nonce + ciphertext).decode()

    def decrypt(self, token: str) -> str:
        raw = base64.b64decode(token)
        nonce, ciphertext = raw[:12], raw[12:]
        return self._aesgcm.decrypt(nonce, ciphertext, None).decode()

File: server/src/lib/crypto.ts
import { env } from "@/env";

const NONCE_LENGTH = 12;

function createCryptoService(base64Key: string) {
  const keyBytes = Buffer.from(base64Key, "base64");
  if (keyBytes.length !== 32) {
    throw new Error(`ENCRYPTION_KEY must be exactly 32 bytes, got ${keyBytes.length}`);
  }

  let cryptoKey: CryptoKey | null = null;

  async function getKey(): Promise<CryptoKey> {
    if (!cryptoKey) {
      cryptoKey = await crypto.subtle.importKey(
        "raw",
        keyBytes,
        "AES-GCM",
        false,
        ["encrypt", "decrypt"],
      );
    }
    return cryptoKey;
  }

  return {
    async encrypt(plaintext: string): Promise<string> {
      const key = await getKey();
      const nonce = crypto.getRandomValues(new Uint8Array(NONCE_LENGTH));
      const encoded = new TextEncoder().encode(plaintext);
      const ciphertext = new Uint8Array(
        await crypto.subtle.encrypt({ name: "AES-GCM", iv: nonce }, key, encoded),
      );
      // nonce(12) + ciphertext + authTag(16)
      const result = new Uint8Array(nonce.length + ciphertext.length);
      result.set(nonce);
      result.set(ciphertext, nonce.length);
      return Buffer.from(result).toString("base64");
    },

    async decrypt(token: string): Promise<string> {
      const key = await getKey();
      const raw = Buffer.from(token, "base64");
      const nonce = raw.subarray(0, NONCE_LENGTH);
      const ciphertext = raw.subarray(NONCE_LENGTH);
      const decrypted = await crypto.subtle.decrypt(
        { name: "AES-GCM", iv: nonce },
        key,
        ciphertext,
      );
      return new TextDecoder().decode(decrypted);
    },
  };
}

export const cryptoService = createCryptoService(env.ENCRYPTION_KEY);
